### PR TITLE
unr-833: Lazy load FClassInfo

### PR DESCRIPTION
**Pull request by tenevdev**
#### Description
ClassInfoMap is populated on-demand when FindClassInfoByClass is called only for the given class.

#### Tests
Reproduced the issue where a level blueprint is not loaded correctly by moving the type-bindings initialization back to where it was. This made a client in https://github.com/spatialos/UnrealGDKThirdPersonShooter crash. Then I made the changes and tested again.

No automated tests.

#### Documentation
Changes are not user-facing.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp [hunk 1]
```diff
@@ -28,169 +28,142 @@ void USpatialTypebindingManager::Init(USpatialNetDriver* InNetDriver)
...
-
-			ActorInfo->SubobjectInfo.Add(Offset, MakeShared<FClassInfo>(SubobjectInfo));
-		}
-	}
+	return Info;
 }
 
 FClassInfo* USpatialTypebindingManager::FindClassInfoByClass(UClass* Class)
 {
+	if (!ClassInfoMap.Contains(Class))
```
Comment (tenevdev) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: I wonder if we should check whether the given class is also in the schema database class paths.

**Comment by tenevdev** [comment]
Also note that callers of USpatialTypebindingManager::FindClassInfoByClass can't use the returned pointer indefinitely because with the existing map implementation it is invalidated by changes to the map.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp (hunk 1 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: Adding a IsSupportedClass() initially, logging the warning, and return nullptr sounds like a good idea.

**Comment by m-samiec** [comment]
Great work! Few comments.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp [hunk 2]
```diff
@@ -28,169 +28,142 @@ void USpatialTypebindingManager::Init(USpatialNetDriver* InNetDriver)
...
+	{
+		FClassInfo Info = CreateTypebindingsForClass(Class);
```
[the remark this answers is not in the document]
Comment (tenevdev) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: I thought return value optimisation would handle it for me.

[the remark this answers is not in the document]
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: RVO will potential assist in the immediate return, but not the copy when adding to the map.

[the remark this answers is not in the document]
Comment (tenevdev) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: Thanks for the clarification :)

**Comment by tenevdev** [comment]
I added another tweak to make sure there is only one lookup - before it did Contains and Find.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 3]
```diff
@@ -48,6 +48,9 @@ bool USpatialNetDriver::InitBase(bool bInitAsClient, FNetworkNotify* InNotify, c
 	bConnectAsClient = bInitAsClient;
 	bAuthoritativeDestruction = true;
 
+	TypebindingManager = NewObject<USpatialTypebindingManager>();
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Is it ok not to check if it's `nullptr`?

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: This is how it used to be, should always be nullptr as this is now back in InitBase()

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp [hunk 4]
```diff
@@ -15,6 +15,8 @@
 #include "EngineClasses/SpatialPackageMapClient.h"
 #include "Utils/RepLayoutUtils.h"
 
+DEFINE_LOG_CATEGORY(LogTypebindingManager);
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: Can this be changed to `LogSpatialTypebindingManager`

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp [hunk 5]
```diff
@@ -28,170 +30,148 @@ void USpatialTypebindingManager::Init(USpatialNetDriver* InNetDriver)
...
 FClassInfo* USpatialTypebindingManager::FindClassInfoByClass(UClass* Class)
 {
-	return ClassInfoMap.Find(Class);
+	if (!IsSupportedClass(Class))
+	{
+		UE_LOG(LogTypebindingManager, Warning, TEXT("Could not find class in schema database: %s"), *Class->GetPathName());
+		return nullptr;
+	}
+
+	if (FClassInfo* Info = ClassInfoMap.Find(Class))
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]:
```suggestion
	if (!ClassInfoMap.Contains(Class))
	{
		AddTypebindingsForClass(Class);
	}
	
	return ClassInfoMap.Find(Class);
```

Comment (tenevdev) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: It was like this in my first implementation, but this does two searches in the map - one for Contains and one for Find.

**Comment by tenevdev** [comment]
Looks like lazy-loading the `ComponentToClassMap` is not trivial - it will probably benefit from some additional information stored during code generation. This is required for `USpatialTypebindingManager::FindClassInfoByComponentId` to work as before.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp (hunk 5 above)
Comment (tenevdev) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: Discussed offline and will accept the suggestion in favour of readability.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp (hunk 4 above)
Comment (tenevdev) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: Done

**Comment by tenevdev** [comment]
Testing with the 3rd-person shooter project I see that some classes are not found in the schema database.

```
LogSpatialTypebindingManager: Warning: Failed to find class at path for component 58 in schema database
LogSpatialTypebindingManager: Warning: Failed to find class at path for component 50 in schema database
LogSpatialTypebindingManager: Warning: Failed to find class at path for component 54 in schema database
LogSpatialTypebindingManager: Warning: Failed to find class at path for component 100001 in schema database
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp [hunk 6]
```diff
@@ -28,169 +30,194 @@ void USpatialTypebindingManager::Init(USpatialNetDriver* InNetDriver)
...
+{
+	return SchemaDatabase->ClassPathToSchema.Contains(Class->GetPathName());
 }
 
 FClassInfo* USpatialTypebindingManager::FindClassInfoByClass(UClass* Class)
 {
+	// This could be optimised to a single map lookup in all cases if we Find first, but we keep this pattern for readability
+	if (!ClassInfoMap.Contains(Class))
+	{
+		return &AddTypebindingsForClass(Class);
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]:
```suggestion
		AddTypebindingsForClass(Class);
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp [hunk 7]
```diff
@@ -28,169 +30,194 @@ void USpatialTypebindingManager::Init(USpatialNetDriver* InNetDriver)
 		FMessageDialog::Debugf(FText::FromString(TEXT("SchemaDatabase not found! No classes will be supported for SpatialOS replication.")));
 		return;
 	}
-
-	FindSupportedClasses();
-	CreateTypebindings();
 }
 
-void USpatialTypebindingManager::FindSupportedClasses()
+FClassInfo& USpatialTypebindingManager::AddTypebindingsForClass(UClass* Class)
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]:
```suggestion
void USpatialTypebindingManager::AddTypebindingsForClass(UClass* Class)
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp [hunk 8]
```diff
@@ -28,169 +30,194 @@ void USpatialTypebindingManager::Init(USpatialNetDriver* InNetDriver)
...
 		});
 
-		Info.Class = Class;
-
-		ClassInfoMap.Emplace(Class, Info);
+		Info.SubobjectInfo.Add(Offset, MakeShared<FClassInfo>(SubobjectInfo));
 	}
 
-	for (UClass* Class : SupportedClasses)
+	return Info;
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialTypebindingManager.h [hunk 9]
```diff
@@ -93,8 +95,8 @@ class SPATIALGDK_API USpatialTypebindingManager : public UObject
 	ESchemaComponentType FindCategoryByComponentId(Worker_ComponentId ComponentId);
 
 private:
-	void FindSupportedClasses();
-	void CreateTypebindings();
+	FClassInfo& AddTypebindingsForClass(UClass* Class);
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialTypebindingManager.h]:
```suggestion
	void AddTypebindingsForClass(UClass* Class);
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp [hunk 10]
```diff
@@ -28,169 +30,194 @@ void USpatialTypebindingManager::Init(USpatialNetDriver* InNetDriver)
...
-			if (Property->PropertyFlags & CPF_Handover)
-			{
-				for (int32 ArrayIdx = 0; ArrayIdx < PropertyIt->ArrayDim; ++ArrayIdx)
-				{
-					FHandoverPropertyInfo HandoverInfo;
-					HandoverInfo.Handle = Info.HandoverProperties.Num() + 1; // 1-based index
-					HandoverInfo.Offset = Property->GetOffset_ForGC() + Property->ElementSize * ArrayIdx;
-					HandoverInfo.ArrayIdx = ArrayIdx;
-					HandoverInfo.Property = Property;
+	Info.Class = Class;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: nit: move this to immediately after being added to the map

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp [hunk 11]
```diff
@@ -260,6 +286,13 @@ bool USpatialTypebindingManager::FindOffsetByComponentId(Worker_ComponentId Comp
 		return true;
 	}
 
+	if (UClass* Class = FindClassByComponentId(ComponentId))
+	{
+		AddTypebindingsForClass(Class);
+		OutOffset = ComponentToOffsetMap[ComponentId];
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: Worth adding a check here that the ComponentToOffsetMap entry exists

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp [hunk 12]
```diff
@@ -241,15 +262,20 @@ FClassInfo* USpatialTypebindingManager::FindClassInfoByObject(UObject* Object)
...
-bool USpatialTypebindingManager::IsSupportedClass(UClass* Class)
+UClass* USpatialTypebindingManager::FindClassByComponentId(Worker_ComponentId ComponentId)
 {
-	return SupportedClasses.Contains(Class);
+	if (UClass** Class = ComponentToClassMap.Find(ComponentId))
+	{
+		return *Class;
+	}
+
+	return LoadClassForComponent(ComponentId);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: Any reason we don't call `AddTypebindingsForClass(Class)` here when a new class is loaded?

Comment (tenevdev) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: It wasn't needed to handle this particular function as it only cares about the class. However, I see your point and I've made the side effects of all Find* functions more consistent by including `AddTypebindingsForClass` calls for the actor class associated to the given component id inside `LoadClassForComponent`.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp (hunk 11 above)
Comment (tenevdev) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: It must exist, otherwise the implementation is broken. Should this be a `FindChecked` or I should stick to `Find` and null-check?

**Comment by tenevdev** [comment]
@improbable-valentyn I tried to extract the common control flow pattern out of the Find functions which operate on component id's as you asked me. I tried using unreal delegates or function pointers to create a map class with lazy-loading functionality but in the end it was too complicated and would impact readability. Let me know if you have any other ideas how to tackle this duplication without adding much complexity to the code.

**Comment by tenevdev** [comment]
When testing the latest changes in a local deployment I get warnings from `FNetworkPredictionData_Client_Character::CreateSavedMove` - "CreateSavedMove: Hit limit of 96 saved moves (timing out or very bad ping?)". What could be causing this?

**Comment by tenevdev** [comment]
I see the same issue only when running 2 servers on the master branch too.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp [hunk 13]
```diff
@@ -28,169 +30,207 @@ void USpatialTypebindingManager::Init(USpatialNetDriver* InNetDriver)
...
+UClass* USpatialTypebindingManager::LoadClassForComponent(Worker_ComponentId ComponentId)
+{
+	// Only try to load classes for generated components
+	if (ComponentId < SpatialConstants::STARTING_GENERATED_COMPONENT_ID)
+	{
+		return nullptr;
 	}
 
-	for (UClass* Class : SupportedClasses)
+	for (auto& ObjectDataPair : SchemaDatabase->ClassPathToSchema)
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: Nit: `ObjectDataPair` sounds like it stores something for an object instance, while this actually stores a pair of class path and class info. `ClassInfoPair`?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp [hunk 14]
```diff
@@ -28,169 +30,207 @@ void USpatialTypebindingManager::Init(USpatialNetDriver* InNetDriver)
...
 			}
+		}
 
-			// Make a copy of the already made FClassInfo for this specific subobject
-			FClassInfo SubobjectInfo = *SubobjectInfoPtr;
-
-			SubobjectInfo.SubobjectName = SubobjectSchemaData.Name;
-
-			ForAllSchemaComponentTypes([&](ESchemaComponentType Type)
+		for (auto& SubobjectDataPair : ObjectDataPair.Value.SubobjectData)
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: To be consistent with my suggestion above, `SubobjectInfoPair`?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp [hunk 15]
```diff
@@ -28,169 +30,207 @@ void USpatialTypebindingManager::Init(USpatialNetDriver* InNetDriver)
...
-				if (ComponentId != 0)
+				const Worker_ComponentId SubobjectComponentId = SubobjectDataPair.Value.SchemaComponents[Type];
+				if (ComponentId == SubobjectComponentId)
 				{
-					SubobjectInfo.SchemaComponents[Type] = ComponentId;
-					ComponentToClassMap.Add(ComponentId, SubobjectClass);
-					ComponentToOffsetMap.Add(ComponentId, Offset);
-					ComponentToCategoryMap.Add(ComponentId, (ESchemaComponentType)Type);
+					FSoftClassPath SoftClassPath(SubobjectDataPair.Value.ClassPath);
+					UClass* Class = SoftClassPath.TryLoadClass<UObject>();
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: Do we need to explicitly load the subobject's class?
Supposedly loading the actor class should also load all its components' classes.
@m-samiec @Vatyx any thoughts?

Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: In fact, this should never need to load the class, only check that it is already loaded and get the class pointer, which you can do with `SoftClassPath.ResolveClass()`. If it returns null, that would indicate a problem with our logic, so we should just `check` the result of `ResolveClass()`.
The reason for this is that we should have already loaded the class and all of the subobjects' classes when creating the actor, which happens before we go into this.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp [hunk 16]
```diff
@@ -28,169 +30,207 @@ void USpatialTypebindingManager::Init(USpatialNetDriver* InNetDriver)
...
 
-		for (TFieldIterator<UProperty> PropertyIt(Class); PropertyIt; ++PropertyIt)
+	ForAllSchemaComponentTypes([&](ESchemaComponentType Type)
+	{
+		Worker_ComponentId ComponentId = SchemaDatabase->ClassPathToSchema[Class->GetPathName()].SchemaComponents[Type];
+		if (ComponentId != 0)
 		{
-			UProperty* Property = *PropertyIt;
+			Info.SchemaComponents[Type] = ComponentId;
+			ComponentToClassInfoMap.Emplace(ComponentId, MakeShared<FClassInfo>(Info));
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: We're duplicating the `FClassInfo` here. What we should do instead is change `ClassInfoMap` to `TMap<UClass*, TSharedRef<FClassInfo>>` and then pass that `TSharedRef` here.
Note the usage of `TSharedRef` instead of `TSharedPtr`, indicating that it can't be null.

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialTypebindingManager.h [hunk 17]
```diff
@@ -71,6 +71,8 @@ struct FClassInfo
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialTypebindingManager.h]: `SubobjectInfo` above should also probably be changed to map to `TSharedRef`.

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialTypebindingManager.h [hunk 18]
```diff
@@ -103,13 +106,10 @@ class SPATIALGDK_API USpatialTypebindingManager : public UObject
...
 	USchemaDatabase* SchemaDatabase;
 
-	UPROPERTY()
-	TArray<UClass*> SupportedClasses;
-
 	UPROPERTY()
 	TMap<UClass*, FClassInfo> ClassInfoMap;
 
-	TMap<Worker_ComponentId, UClass*> ComponentToClassMap;
+	TMap<Worker_ComponentId, TSharedPtr<FClassInfo>> ComponentToClassInfoMap;
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialTypebindingManager.h]:
```suggestion
	TMap<Worker_ComponentId, TSharedRef<FClassInfo>> ComponentToClassInfoMap;
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp (hunk 13 above)
Comment (tenevdev) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: I believe this is the same naming as in the previous implementation. Note that value of the pair is not class info, it's a `FSchemaData`. Do you still think it should be changed?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp (hunk 15 above)
Comment (tenevdev) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: I agree the sub-object class doesn't need to be loaded with the current implementation as it would be loaded as part of the call to `AddTypebindingsForClass(ActorClass)` a few lines below.

Regarding your second point, I'm not familiar with exactly how the actors are created, but without assuming anything it seems possible to me that the sub-object class is not already loaded at this point and will in fact be loaded in a call to  `AddTypebindingsForClass`.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp (hunk 13 above)
Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: You're right. I would prefer `ClassDataPair` though (my problem is that `Object` seems to imply an object instance).

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp (hunk 15 above)
Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: I might be wrong, but my current understanding is that we don't even need to load the class in `AddTypebindingsForClass`, only try to resolve it. The reasoning is that the entity would have to have been checked out before we could receive updates or commands on it, which means that we would need to spawn the corresponding actor first, which means that the actor class as well as the subobject classes would be already loaded.

Comment (tenevdev) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: Tested this and it seems to work ok.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp (hunk 16 above)
Comment (tenevdev) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: Follow-up task https://improbableio.atlassian.net/browse/UNR-930 . Hopefully, I get a chance to do it.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 19]
```diff
@@ -636,6 +636,7 @@ void USpatialReceiver::OnComponentUpdate(Worker_ComponentUpdateOp& Op)
 	case SpatialConstants::METADATA_COMPONENT_ID:
 	case SpatialConstants::POSITION_COMPONENT_ID:
 	case SpatialConstants::PERSISTENCE_COMPONENT_ID:
+	case SpatialConstants::INTEREST_COMPONENT_ID:
```
Comment (danielimprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Nice driveby:)

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp [hunk 20]
```diff
@@ -28,169 +30,208 @@ void USpatialTypebindingManager::Init(USpatialNetDriver* InNetDriver)
...
-				continue;
-			}
-
-			FClassInfo* SubobjectInfoPtr = FindClassInfoByClass(SubobjectClass);
-			if (SubobjectInfoPtr == nullptr)
+			const Worker_ComponentId ObjectComponentId = ClassDataPair.Value.SchemaComponents[Type];
+			if (ComponentId == ObjectComponentId)
 			{
-				continue;
+				FSoftClassPath SoftClassPath(ClassDataPair.Key);
```
Comment (danielimprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: This code could be lifted out into a helper function as the same pattern is used in three places.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp [hunk 21]
```diff
@@ -28,169 +30,200 @@ void USpatialTypebindingManager::Init(USpatialNetDriver* InNetDriver)
...
-void USpatialTypebindingManager::FindSupportedClasses()
+FORCEINLINE UClass* ResolveClass(FString& ClassPath)
 {
-	TArray<FString> SupportedClassPaths;
-	SchemaDatabase->ClassPathToSchema.GetKeys(SupportedClassPaths);
-
-	for (auto& ClassPath : SupportedClassPaths)
+	FSoftClassPath SoftClassPath(ClassPath);
+	UClass* Class = SoftClassPath.ResolveClass();
+	if (Class == nullptr)
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: Replace this with
```
checkf(Class, TEXT("Failed to load class at path %s"), *ClassPath);
```
The reason is that not resolving the class here is fatal, and we will crash because of it (and it's not worth guarding everywhere we call this)

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp [hunk 22]
```diff
@@ -28,169 +30,200 @@ void USpatialTypebindingManager::Init(USpatialNetDriver* InNetDriver)
...
-	// but schema components will be 0 since subobject classes don't have associated components
-	// afterwards, the subobject FClassInfos will be the basis of the FClassInfo for specific subobjects
-	// and have their proper schema components filled in
-	for (UClass* Class : SupportedClasses)
-	{
-		FClassInfo Info;
+	checkf(IsSupportedClass(Class), TEXT("Could not find class in schema database: %s"), *Class->GetPathName());
 
-		TArray<UFunction*> RelevantClassFunctions = improbable::GetClassRPCFunctions(Class);
+	FClassInfo& Info = ClassInfoMap.Add(Class);
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp [hunk 23]
```diff
@@ -28,169 +30,191 @@ void USpatialTypebindingManager::Init(USpatialNetDriver* InNetDriver)
...
+UClass* USpatialTypebindingManager::LoadClassForComponent(Worker_ComponentId ComponentId)
+{
+	// Only try to load classes for generated components
+	if (ComponentId < SpatialConstants::STARTING_GENERATED_COMPONENT_ID)
+	{
+		return nullptr;
 	}
 
-	for (UClass* Class : SupportedClasses)
+	for (auto& ClassDataPair : SchemaDatabase->ClassPathToSchema)
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: Maybe I'm just noticing this now, but the runtime of this function is fairly bad. I think it might be worth creating a map ahead of time from component Id -> class path and using that for look ups.

Comment (tenevdev) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: This could be done in schema generation and will probably help if performance is an issue.

Comment (Vatyx) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialTypebindingManager.cpp]: The map is going to be as large as any other maps containing component ids would be. I think it's potentially worth doing now since I don't like that we are iterating so much every time we load a new class.
